lib/test.js

```diff
@@ -365,21 +365,19 @@ class Test {
 	}
 
 	finish() {
-		this.finishing = true;
-
 		if (!this.assertError && this.pendingThrowsAssertion) {
 			return this.waitForPendingThrowsAssertion();
 		}
 
-		this.verifyPlan();
-		this.verifyAssertions();
-
 		if (this.assertError || this.pendingAssertions.length === 0) {
 			return this.completeFinish();
 		}
 
-		// Finish after potential errors from pending assertions have been consumed.
-		return Promise.all(this.pendingAssertions).then(() => this.completeFinish());
+		// Retry finishing after potential errors from pending assertions have been consumed.
+		const consumedErrors = Promise.all(this.pendingAssertions);
+		// Note that more assertions may be added in the meantime.
+		this.pendingAssertions = [];
```

Comment:
Maybe add a test to validate the works it happens twice (pending resolve, but more pending have been added, and yet another sync will be added while you wait for those to resolve).